### PR TITLE
Breaking: acorn 8.0.5

### PR DESCRIPTION
**Pull request by mreinstein**

### CODE REVIEW

**Comment by nzakas** [comment]
Thanks. Can you summarize the changes in the latest version of Acorn? (We want to make sure there are no breaking changes that would flow through to Espree.)

**Comment by mreinstein** [comment]
> Can you summarize the changes in the latest version of Acorn? 

Sure! Where is the appropriate place to summarize these changes?

**Comment by mreinstein** [comment]
The only breaking change in acorn@8.x is that `ecmaVersion` is now a required field.

**Comment by mreinstein** [comment]
https://github.com/acornjs/acorn/blob/master/acorn/CHANGELOG.md#800-2020-08-12

**Comment by mdjermanovic** [comment]
It would be probably good to also upgrade `acorn-jsx` to the latest version, as v5.2.0 doesn't have acorn v8 as a peer dependency: 

https://github.com/acornjs/acorn-jsx/blob/c30617bd8d3763ee96fc76abfc0a9bb00e036d68/package.json#L22

**Comment by mdjermanovic** [comment]
It seems that this change wasn't included in the changelog:

https://github.com/acornjs/acorn/commit/85f41501c536d9141e50f616f5db6b37e5f0b31b

An example of the difference in AST:

```js
const acorn = require("acorn");

const code = "({x, y = 1} = {z})";

const ast = acorn.parse(code, { ecmaVersion: 2015 });

const assignment = ast.body[0].expression;
const objectPattern = assignment.left;
const objectExpression = assignment.right;

let propertyX = objectPattern.properties[0];
let propertyY = objectPattern.properties[1];
let propertyZ = objectExpression.properties[0];

console.log(propertyX.key === propertyX.value); // true in v7, false in v8
console.log(propertyY.key === propertyY.value.left); // true in v7, false in v8
console.log(propertyZ.key === propertyZ.value); // true in v7, false in v8

```

This change was released in `v8.0.0`. I'm not sure if it constitutes a breaking change in Acorn, but wouldn't be surprised if some ESLint rules need to be modified a bit. Actually, we already have one issue waiting for this: https://github.com/eslint/eslint/issues/13021

**Comment by mreinstein** [comment]
@mdjermanovic wow, nice catch on that change that slipped through the log! What should we do? Should we just stay on acorn@7?

**Comment by mdjermanovic** [comment]
> Should we just stay on acorn@7?

We'll have to upgrade Acorn sooner or later for new syntax, so it might be good to start with that now.

I believe https://github.com/acornjs/acorn/commit/85f41501c536d9141e50f616f5db6b37e5f0b31b should be a good change for ESLint rules. If some code relies (intentionally or not) on previous behavior, then it probably doesn't work well with `@babel/eslint-parser` and `@typescript-eslint/parser,` since they already produce AST like acorn v8 in the above examples (multiple `Identifier` objects instead of reusing the same instance for shorthand properties).

Also, without this change `parent` property seems broken for `({ a = 1 } = {})` as described in https://github.com/eslint/eslint/issues/13021.

We could fix what has to be fixed in ESLint before releasing `espree`, and then release both.

Problems might be with:

* Plugins and custom rules that might need to be fixed as well. 
* Tools that use `espree`, I'm not sure what should be the next `espree` version after this change.

@nzakas what do you think about this?

**Comment by nzakas** [comment]
That sounds like a good plan. To be safe, I think I’d do a major version bump for Espree. While ESLint is the main consumer, we don’t really know where else it might be used. Even though it seems like a bug fix, I’d err on the side of caution since we don’t pin dependencies in ESLint, new installs might just break if we do this as a minor bump.

**Comment by mdjermanovic** [comment]
:+1: for Espree major version.

**Comment by nzakas** [comment]
We should probably coordinate on a major release to see if we want to do #457 as well

**Comment by mreinstein** [comment]
in related news: https://twitter.com/MylesBorins/status/1329163754639155201

Sounds like the node team is about to mark the es module stable, in both 12.x and 14.x release lines.

depending on your willingness to drop node support below node v12.17, it might just make sense to avoid the dual package approach, switch to a es module only, and make espree 8.x node v12.17 and up.

**Comment by mdjermanovic** [comment]
> It would be probably good to also upgrade `acorn-jsx` to the latest version, as v5.2.0 doesn't have acorn v8 as a peer dependency

Prepared #459 to upgrade `acorn-jsx`, and a couple of test for default `ecmaVersion` in #460.

Should we add some tests for [this change](https://github.com/eslint/espree/pull/456#issuecomment-729299535)?

**Comment by nzakas** [comment]
@mreinstein i don’t think we can do just a hard cutover to only an ESM package. We just don’t know who else is depending on this package and we don’t want to cut people off unnecessarily.

@mdjermanovic i think tests would be a good idea. 👍

**Comment by mdjermanovic** [comment]
@mreinstein can you add tests for [this change](https://github.com/eslint/espree/pull/456#issuecomment-729299535)?

I think [tests/lib/parse.js](https://github.com/eslint/espree/blob/master/tests/lib/parse.js) would be the right place.

**Comment by mreinstein** [comment]
@mdjermanovic I can add tests (and via a clean PR) if we want this approach.

On a related note, concerning acorn bump and es module:

If we are going to bump the major version because we're concerned about possibly breaking behavior coming from acorn@8, I would advocate for putting the dual package into the next major release too. (#458 )

As an end user having espree 8 release with a potential behavior change of acorn, and then having to go through another potentially painful upgrade when espree 9 includes a dual package is a bit stressful.

**Comment by nzakas** [comment]
@mreinstein yup, that’s what I was advocating for [here](https://github.com/eslint/espree/pull/456#issuecomment-732534843).

**Comment by mreinstein** [comment]
I'll have the PR updated for this tomorrow, I'm just occupied with an American holiday today (thanksgiving.) Sorry for the delay!

**Comment by nzakas** [comment]
No need to apologize for taking down time. We're not in any hurry.

**Comment by mreinstein** [comment]
> We're not in any hurry.

I know there's no hurry, and I also know this is not in the typical github contribution culture, but I like to provide an ETA when I feel confident that I can meet it. :-) 

Anyway, I've added a test as discussed. I put it in `acorn-after-espree.js` rather than `parse.js`; figured this matches the other underlying acorn tests, but if we want to move it to `parse.js` I'm happy to.

File: tests/lib/acorn-after-espree.js [hunk 1]
```diff
@@ -19,11 +19,30 @@ const acorn = require("acorn"),
...
-        const after = acorn.parse("var foo = bar /*world*/;");
+        const after = acorn.parse("var foo = bar /*world*/;", { ecmaVersion: 5 });
 
         assert.deepStrictEqual(after, before);
     });
+
+    it("acorn.parse() should not re-use the same identifier node in object patterns", () => {
+        const code = "({x, y = 1} = {z})";
+
+        const ast = acorn.parse(code, { ecmaVersion: 2015 });
```
Comment (mdjermanovic) [inline tests/lib/acorn-after-espree.js:33]: We should test what espree is producing at the end, not acorn directly.

If acorn reverts this change in the future for some reason, we could then clone the nodes in espree to retain the behavior.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: tests/lib/parse.js [hunk 2]
```diff
@@ -82,4 +82,24 @@ describe("parse()", () => {
         });
 
     });
+
+    describe("acorn", () => {
+        it("acorn.parse() should not re-use the same identifier node in object patterns", () => {
```
Comment (mdjermanovic) [inline tests/lib/parse.js:87]:
```suggestion
        it("should not re-use the same identifier node in shorthand properties", () => {
```

since it's also testing object expressions

File: tests/lib/parse.js [hunk 3]
```diff
@@ -82,4 +82,24 @@ describe("parse()", () => {
...
+            const objectPattern = assignment.left;
+            const objectExpression = assignment.right;
+
+            const propertyX = objectPattern.properties[0];
+            const propertyY = objectPattern.properties[1];
+            const propertyZ = objectExpression.properties[0];
+
+            assert.notStrictEqual(propertyX.key, propertyX.value); // true in v7, false in v8
+            assert.notStrictEqual(propertyY.key, propertyY.value.left); // true in v7, false in v8
+            assert.notStrictEqual(propertyZ.key, propertyZ.value); // true in v7, false in v8
```
Comment (mdjermanovic) [inline tests/lib/parse.js:102]:
```suggestion
            assert.notStrictEqual(propertyX.key, propertyX.value);
            assert.notStrictEqual(propertyY.key, propertyY.value.left);
            assert.notStrictEqual(propertyZ.key, propertyZ.value);
```

I think there's no need for the version comments here.

**Review by mdjermanovic — COMMENTED** [review]
The test assertions look good! I left some notes about the descriptions and comments.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

File: package.json [hunk 4]
```diff
@@ -18,7 +18,7 @@
   },
   "license": "BSD-2-Clause",
   "dependencies": {
-    "acorn": "^7.4.0",
+    "acorn": "^8.0.4",
```
Comment (mdjermanovic) [inline package.json:21]:
```suggestion
    "acorn": "^8.0.5",
```

Can you update this, so we don't forget. 

v8.0.4 doesn't work on some Node versions we support (fixed by https://github.com/acornjs/acorn/commit/0a4167509adee401490464f5d2ff3eeefc2cf341).

**Review by mdjermanovic — APPROVED** [review]
Thanks!

**Comment by mreinstein** [comment]
closing in favor of #469
